### PR TITLE
sync: avoid creating mutex and rwlock resource_span with parent, use None parent instead

### PR DESCRIPTION
**Pull request by xuorig**
Most of the context for this PR can be found in [this issue](https://github.com/tokio-rs/tokio/issues/6106).

The `TRACE` `resource_span` that `Mutex` and `RWLock` create when using `new` can keep open their parent for as long as the mutex lives. This pull request changes these spans to have a parent of `None` rather than having a *parent-child* relation ship with the current span.

As mentioned by @davidbarsky in the linked issue, another possibility could be to alter the API and accept a parent span. I'm open to changing this PR although I agree it feels strange when it comes to `tracing`.

Looking for existing tests, I wasn't sure if regression testing this (which would require bringing a tracing subscriber most likely) was necessary here. Open to adding this as well.

### CODE REVIEW

**Comment by NobodyXu** [comment]
OnceCell, Semaphore etc probably also have the same issue, IMO it's a good idea to also change that in this PR

**Comment by Darksonn** [comment]
Yeah, if possible I'd like to fix all instances of this.

I don't think we have many tests for the tracing stuff. 

> Open to adding this as well.

If you want to add tests, then those are always welcome.

File: tokio/src/sync/mutex.rs
Comment (davidbarsky) [inline tokio/src/sync/mutex.rs:342]: This change is necessary, but—IMO—not sufficient. I think we should probably also add a `follows_from` relationship to any resource spans (in contradiction of https://github.com/tokio-rs/tokio/issues/5792) since it doesn't have any impact on memory, leaks or things of that nature. I can't make the following a suggestion because the GitHub only allows me to leave suggestions on _what changed_, but please treat the following as a suggestion:

```rust
            let span = tracing::trace_span!(
                parent: None,
                "runtime.resource",
                concrete_type = "Mutex",
                kind = "Sync",
                loc.file = location.file(),
                loc.line = location.line(),
                loc.col = location.column(),
            );
            span.follows_from(Span::current());
            span
```

(note that [this](https://docs.rs/tracing/latest/tracing/struct.Span.html#method.current) is `Span::current()`. You might need to import it.)

File: tokio/src/sync/rwlock.rs
Comment (davidbarsky) [inline tokio/src/sync/rwlock.rs:212]: similar change 🙂:
```rust
        let span = tracing::trace_span!(
                parent: None,
                "runtime.resource",
                concrete_type = "RwLock",
                kind = "Sync",
                loc.file = location.file(),
                loc.line = location.line(),
                loc.col = location.column(),
            );
            span.follows_from(Span::current());
            span
```

**Review by davidbarsky — CHANGES_REQUESTED** [review]
I agree with Alice and Jiahao–can you make changes to `tokio::sync:OnceCell`, `tokio::sync::Semaphore`, and `tokio::sync::Barrier` similar to my suggestions on `tokio::sync::RwLock`/`tokio::sync::Mutex` ?

**Comment by davidbarsky** [comment]
@xuorig On the subject of tests, I know that @hds was looking at adding tests for resources in Tokio using [`tracing-mock`](https://github.com/tokio-rs/tracing/tree/v0.1.x/tracing-mock). What are your timelines for needing this to land? Would y'all be able to take a dependency on an unmerged PR to Tokio for a few days or nah?

File: tokio/src/sync/mutex.rs
Comment (xuorig) [reply tokio/src/sync/mutex.rs:342]: That makes sense, right now it would be a no root span in the "void". I'll make that change, thank you.

**Comment by xuorig** [comment]
@davidbarsky absolutely, this can wait. I've added a mock currently but happy to swap it out later 👍

File: tokio/src/sync/mutex.rs
Comment (xuorig) [reply tokio/src/sync/mutex.rs:342]: Oh and also, I think it would be more something like this:

```rust
           let current_span = tracing::Span::current();
            let span = tracing::trace_span!(
                parent: None,
                "runtime.resource",
                concrete_type = "Mutex",
                kind = "Sync",
                loc.file = location.file(),
                loc.line = location.line(),
                loc.col = location.column(),
            );
            span.follows_from(current_span);
            span
```

Since `Span::current()` called there will be equal to the trace span we just created.

Comment (xuorig) [reply tokio/src/sync/mutex.rs]: Confirmed through [this test](https://github.com/tokio-rs/tokio/pull/6107/files#diff-47b62efd1013ab2926c772c27e5918a5ce58e8f2e65e0821f2050194467b2a83R193). It does require a fairly large setup, so I'm happy to wait for @hds's`tracing-mock` integration, depending on what you folks would prefer.

**Comment by xuorig** [comment]
I just committed a test using tracing-mock before realizing [this PR](https://github.com/tokio-rs/tokio/pull/6112) is introducing it more generally. Will wait for it to be merged and modify this PR.

File: tokio/src/sync/mutex.rs
Comment (davidbarsky) [reply tokio/src/sync/mutex.rs]: Ah, I assumed that that the current span wouldn't change until it was entered. whoops.

**Comment by Darksonn** [comment]
The other PR is merged now.

**Review by hds — CHANGES_REQUESTED** [review]
After some prototyping with tokio-console (the console-subscriber specifically) and discussions with @hawkw and @davidbarsky, I believe we should not add `follows_from` relationship to the resource span (and for the same reasons, I don't plan to add it to the task span either).

TLDR: I propose that this change should only add `parent: None` to these resource spans and make no other changes.

The reason for this is a bit long, but I felt it made sense to write it out so that the justification is available when people come looking at this change.

The reason is that using `follows_from` with the current span is a bit delicate. You're using the **global** current span, but specific tracing-subscriber `Layer`s may not be recording that span (it could be filtered out) and this results in the layer never receiving a call to `on_follows_from`.

Here is an example (I'm illustrating based on task spans, but the same is the case for the sleep resource):

```rust
async fn spawn_parent() {
    tokio::task::Builder::new()
        .name("parent_ends")
        .spawn(parent_spawns_child())
        .unwrap();
}

#[tracing::instrument]
async fn parent_spawns_child() {
    tokio::task::Builder::new()
        .name("child_continues")
        .spawn(async {
            loop {
                tokio::time::sleep(Duration::from_secs(100)).await;
            }
        })
        .unwrap();

    tokio::time::sleep(Duration::from_secs(5)).await;
}
```

Because of the `#[tracing::instrument]` on `parent_spawns_child()`, the span hierarchy looks like this:

```
- runtime.spawn[task.name=parent_ends] (id=1)
  - parent_spawns_child (id=274877906945)
    - runtime.spawn[task.name=child_continues] (id=274877906946)
```

And that `parent_spans_child` span in the middle means that a `Layer` which is only looking for tokio instrumentation (i.e. console-subscriber) will never receive a call to `on_follow_from`. If I take the `#[tracing::instrument]` out, then it will.

This means that the console-subscriber (or any subscriber layer) won't be able to rely on `follows_from` unless it registers for all spans which are active. On the other side, the behavior which we proposed here (providing a "weak" connection to the context within which a resource or task is created) can be safely done within the subscriber layer itself, as the layer has the current span stack from its point of view (containing only spans that it is aware of) and in the example above, the console-subscriber would not see the `parent_spans_child` span and could itself keep a connection between the two spans that won't cause the parent span to remain open longer than it should.

**Comment by xuorig** [comment]
PR updated:

- Using parent None and no `follows_from` relation ship as suggest by @hds.
- Since `batch_semaphore` is only used internally and also since it does seem to have a parent-child relationship with its callers, I've omitted `parent: None` from that resource span.
- Tests added for all `new` functions of the sync primitives in `sync`.

**Review by davidbarsky — APPROVED** [review]
lgtm, but i'll let @hds confirm that it's correct.

**Review by hds — APPROVED** [review]
Looks good to me! Thanks for your work on this change!

I will just note here that the tests (which are great and something very valuable to add), don't test the change, which is to say they're not testing that the resource spans are being created as explicit roots (`parent: None`). However, there's actually a defect in tracing-mock which means that you can't test for that right now anyway (tokio-rs/tracing#2440).

Once that defect gets fixed, these tests can be updated, but for now we have significantly more test coverage than before, so let's go ahead and merge this change!

**Comment by xuorig** [comment]
@hds one question about this, just for my own learnings:

> don't test the change, which is to say they're not testing that the resource spans are being created as explicit roots

I was under the impression they did:

```rust
.new_span(mutex_span.clone().with_explicit_parent(None))
```

Now I understand there's a defect with the above, but are you saying this wasn't the way to do it in the first place?

**Pull request merged**
